WIP: GCC with separated runtime libraries

alexfmpe opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change
Opening bare bones PR to collaborate on #132340 
cc @Ericson2314 @sternenseemann 

###### Things done

All cross compilation from `x86_64-unknown-linux-{gnu,musl}`. Plug in the config in `--crossSystem "{ config = \"$config\"; useGccNg = true; }"`

- [x] `buildPackages.gccNgPackages.gcc-unwrapped`
  - [x] draft package committed
  - [x] `aarch64-unknown-linux-gnu`
  - [x] `aarch64-unknown-linux-musl` 
- [x] `gccNgPackages.libgcc`
  - [x] draft package committed
  - [x] `aarch64-unknown-linux-gnu`
  - [x] `aarch64-unknown-linux-musl` 
- [x] `libcCross`
  - [x] draft package committed
  - [ ] `aarch64-unknown-linux-gnu`
  - [x] `aarch64-unknown-linux-musl`
- [ ] `gccNgPackages.libatomic`
  - [x] draft package committed
  - [ ] `aarch64-unknown-linux-gnu`
  - [x] `aarch64-unknown-linux-musl`
- [ ] `gccNgPackages.libstdcxx`
  - [x] draft package committed
  - [ ] `aarch64-unknown-linux-gnu`
  - [ ] `aarch64-unknown-linux-musl`  
    - Linking errors due to messed up relocations, evidentally because of mixing `-fPIC` and `-fno-PIC`

-------

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

###### Upstream patches

- Prefixed binaries in `-B` dirs
  1. https://gcc.gnu.org/pipermail/gcc-patches/2021-August/576725.html
  2. https://gcc.gnu.org/pipermail/gcc-patches/2021-August/577639.html
  3. https://gcc.gnu.org/pipermail/gcc-patches/2021-August/577640.html
  4. https://gcc.gnu.org/pipermail/gcc-patches/2021-August/577638.html

- Thread model specified per lib
  1. https://gcc.gnu.org/pipermail/gcc-patches/2021-August/577880.html

Ericson2314 committed:
MAX WIP: gccNgPackages_11: init

Taking a bunch of stuff from 11 packages

sternenseemann committed:
gccNgPackages_11: make evaluate to a certain extent, flesh out

* `builtins.attrNames gccNgPackages_11` now evaluates

* Added missing inputs

* Removed a bunch of clang-related things

  - clang-only packages: compiler-rt, libcxx, libcxxabi, openmp

  - useLLVM logic

  - stdenvs / cc-wrappers which are not applicable anymore

* Removed LLVM-specific relics like the default callPackage bindings,
  clean up default callPackage scopes for libraries and tools

* Add runtime libraries (may forgot some) to libraries set

sternenseemann committed:
gccNgPackages_11.libstdcxx: fix path

sternenseemann committed:
gccNgPackages_11.libstdcxx: move patches from old location

Ericson2314 committed:
WIP

Ericson2314 committed:
WIP, Can build gcc but not libgcc

Ericson2314 committed:
WIP improve compiler, get farther with libgcc

SuperSandro2000 commented on pkgs/development/compilers/gcc-ng/11/default.nix:
```diff
@@ -0,0 +1,91 @@
+  gcc_meta = {
+    maintainers = with lib.maintainers; [ ericson231 sternenseemann ];
+  };
+  gcc_libs_meta = gcc_meta // {
+    license     = lib.licenses.lgpl3Plus;
+    platforms   = lib.platforms.all;
+  };
+  gcc_tools_meta = gcc_meta // {
+    license     = lib.licenses.gpl3Plus;
+    platforms   = lib.platforms.unix;
+  };
```
```suggestion
  gcc_libs_meta = gcc_meta // {
    license     = licenses.lgpl3Plus;
    platforms   = platforms.all;
  };
  gcc_tools_meta = gcc_meta // with lib; {
    license     = licenses.gpl3Plus;
    platforms   = platforms.unix;
  };
```

Ericson2314 commented:
Oh no, I forgot to push my stuff! I'll mege.

Ericson2314 committed:
Merge branch 'gcc-ng' of github.com:alexfmpe/nixpkgs into gcc-ng

SuperSandro2000 commented on pkgs/development/compilers/gcc-ng/11/gcc/default.nix:
```diff
@@ -0,0 +1,207 @@
+, reproducibleBuild ? true
+, profiledCompiler ? false
+, langJit ? false
+, texinfo ? null
+, perl ? null # lib.optional, for texi2pod (then pod2man)
+, gmp, mpfr, libmpc, gettext, which, patchelf
+, libelf                      # lib.optional, for link-time optimizations (LTO)
+, isl ? null
+, zlib ? null
+, name ? "gcc"
+, gnused ? null
+, nonRelease ? false, flex ? null, bison ? null # Convenience for working on GCC
```
Please stop adding new occurrences of ``? null``. If the feature flag is deactivated the package is no evaluated and does not need to be overwritten to null.

sternenseemann replied:
This derivation may be called from a bootstrap scope where these libraries are missing.

Ericson2314 committed:
gccNgPackages_11.libgcc: Fix build

Create patch so that thread model can be specified explicitly rather
than being inferred from GCC.

Ericson2314 committed:
Merge remote-tracking branch 'upstream/master' into gcc-ng

Ericson2314 committed:
Merge remote-tracking branch 'upstream/master' into gcc-ng

Ericson2314 committed:
gccNgPackages.libatomic: Init

Ericson2314 committed:
Merge remote-tracking branch 'upstream/master' into gcc-ng

Ericson2314 committed:
gccPackagesNg.libgcc: Fix Musl build

Ericson2314 committed:
glibc: Fix evaluation when using gccNg

There is a still a build failure, but I am going to just use Musl and
ignore this for now.

Ericson2314 committed:
libatomic: Fix build

Ericson2314 committed:
gccNgPackages.{libatomic, libstdcxx}: Simplify and improve

Unfortunately, libstdcxx doesn't link, evidentally because inconsistent
`-fPIC`.

Ericson2314 committed:
libssp: Init

SuperSandro2000 commented on pkgs/development/compilers/gcc-ng/11/libatomic/default.nix:
```diff
@@ -0,0 +1,54 @@
+    mkdir -p "$out/lib" "$dev/include"
+    ln -s "$out/lib/gcc/${stdenv.hostPlatform.config}/${version}"/* "$out/lib"
+    ln -s "$dev/lib/gcc/${stdenv.hostPlatform.config}/${version}/include"/* "$dev/include/"
+  '';
```
Meta is missing.

Ericson2314 committed:
gccNgPackages.lib{atomic,ssp}: Add Meta

alexfmpe commented:
Unmark stale

amjoseph-nixpkgs commented:
Hi,

It is a bummer that this sort of fell by the wayside.  I just wanted to mention that enabling the goals of this PR is one of the major benefits sought by #209870.

Specifically, if that PR merges, my next project will be to start moving the libraries that are bundled with gcc into separate packages -- like this PR, but one at a time, and incrementally rather than in a big rewrite.  I would like to reuse as much of the work here as possible.

Ericson2314 commented:
https://github.com/Ericson2314/nixpkgs/tree/gcc-ng-12 This was me trying to update this to a new GCC.

Ericson2314 commented:
@amjoseph-nixpkgs 

https://github.com/NixOS/nixpkgs/pull/132343#issuecomment-1438194839 it is a bummer, and I would love to pair with you on this to get it over the hill at some point. there was just some things I was banging my head against the wall debugging that I need another pair of eyes on.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/2023-10-31-nixpkgs-architecture-team-meeting-45/34846/1

Ericson2314 commented:
Replaced by #414299
